internal/apidiff/apidiff.go

```diff
@@ -0,0 +1,284 @@
+		// Only diff stable clients.
+		if entry.ReleaseLevel != "ga" {
+			continue
+		}
+
+		// Prepare module directory paths relative to the repo root.
+		pkg := strings.TrimPrefix(imp, rootMod+"/")
+		baseModDir := goCloud
+		modDir := d.root
+
+		// Manual clients are also submodules, so we need to run apidiff in the
+		// submodule.
+		if entry.ClientLibraryType == "manual" {
+			baseModDir = path.Join(baseModDir, pkg)
+			modDir = path.Join(modDir, pkg)
+		}
+
+		// Create apidiff base from repo remote HEAD.
+		base, err := d.writeBase(baseModDir, imp, pkg)
+		if err != nil {
+			d.errs[imp] = err
+			continue
+		}
+
+		// Diff the current checked out change against remote HEAD base.
+		out, err := d.diff(modDir, imp, pkg, base)
+		if err != nil {
+			d.errs[imp] = err
+			continue
+		}
+
+		if out != "" {
+			d.results[imp] = out
+		}
+	}
+
+	return nil
+}
+
+func (d *differ) writeBase(baseModDir, imp, pkg string) (string, error) {
+	if err := cd(baseModDir); err != nil {
+		return "", err
+	}
+
+	base := path.Join(baseModDir, "pkg.master")
+	out, err := execE("apidiff", "-w", base, imp)
+	if err != nil && !isSubModErr(out) {
+		return "", err
+	}
+
+	// If there was an issue with loading a submodule, change into that
+	// submodule directory and try again.
+	if isSubModErr(out) {
+		parent := d.manualParent(imp)
+		if parent == pkg {
+			return "", fmt.Errorf("unable to find parent module for %q", imp)
+		}
+		if err := cd(parent); err != nil {
+			return "", err
+		}
+		out, err := execE("apidiff", "-w", base, imp)
+		if err != nil {
+			return "", fmt.Errorf("%s: %s", err, out)
+		}
+	}
+	return base, nil
+}
+
+func (d *differ) diff(modDir, imp, pkg, base string) (string, error) {
+	if err := cd(modDir); err != nil {
+		return "", err
+	}
+	out, err := execE("apidiff", "-incompatible", base, imp)
+	if err != nil && !isSubModErr(out) {
+		return "", err
+	}
+	if isSubModErr(out) {
+		parent := d.manualParent(imp)
+		if parent == pkg {
+			return "", fmt.Errorf("unable to find parent module for %q", imp)
+		}
+		if err := cd(parent); err != nil {
+			return "", err
+		}
+		out, err = execE("apidiff", "-w", base, imp)
+		if err != nil {
+			return "", fmt.Errorf("%s: %s", err, out)
+		}
+	}
+
+	return out, err
+}
+
+func (d *differ) loadManifest() error {
+	f, err := os.Open(repoMetadataPath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	return json.NewDecoder(f).Decode(&d.manifest)
+}
+
+func (d *differ) manualParent(imp string) string {
+	pkg := strings.TrimPrefix(imp, rootMod)
+	split := strings.Split(pkg, "/")
+
+	mod := rootMod
+	for _, seg := range split {
+		mod = path.Join(mod, seg)
+		if parent, ok := d.manifest[mod]; ok && parent.ClientLibraryType == "manual" {
+			return strings.TrimPrefix(mod, rootMod+"/")
+		}
+	}
+
+	return pkg
+}
+
+func shouldRun() (bool, error) {
+	head, err := execE("git", "log", "-1")
+	if err != nil {
+		return false, err
+	}
+	return !strings.Contains(head, "BREAKING_CHANGE_ACCEPTABLE"), nil
```

Comment:
@codyoss can I change this to check that the commit either contains `BREAKING_CHANGE`  _or_ starts with `!`? That way the conventional commit practices used in today's commits will _just work_, especially on go-genproto.

Reply:
Breaking changes should be rare and difficult to make. I'm OK with requiring `BREAKING_CHANGE` in the body.

Reply:
It's actually required by those publishing proto changes to googleapis today. They must either provide a `BREAKING_CHANGE` tag in the body or start their commit with `!` e.g. `!fix: remove accidentally published field`. So we'd be piggy-backing on the existing practice.